packages/sync/src/modules/Full_Sync.php

```diff
@@ -177,75 +193,99 @@ public function start( $module_configs = null ) {
 		$full_sync_queue          = new Queue( 'full_sync' );
 
-		$available_queue_slots = $max_queue_size_full_sync - $full_sync_queue->size();
+		return $max_queue_size_full_sync - $full_sync_queue->size();
+	}
 
-		if ( $available_queue_slots <= 0 ) {
-			return;
-		} else {
-			$remaining_items_to_enqueue = min( Settings::get_setting( 'max_enqueue_full_sync' ), $available_queue_slots );
-		}
+	/**
+	 * Get Modules that are configured to Full Sync and haven't finished enqueuing
+	 *
+	 * @param array $configs Full sync configuration for all sync modules.
+	 *
+	 * @return array
+	 */
+	public function get_remaining_modules_to_enqueue( $configs ) {
+		return array_filter(
+			Modules::get_modules(),
+			function ( $module ) use ( $configs ) {
+				// Skip module if not configured for this sync or module is done.
+				if ( ! isset( $configs[ $module->name() ] ) ) {
+					return false;
+				}
+				if ( ! $configs[ $module->name() ] ) {
+					return false;
+				}
+				if ( ! isset( $this->enqueue_status[ $module->name() ] ) ) {
+					return false;
+				}
+				if ( true === $this->enqueue_status[ $module->name() ][2] ) {
+					return false;
+				}
+
+				return true;
+			}
+		);
+	}
 
+	/**
+	 * Enqueue the next items to sync.
+	 *
+	 * @access public
+	 *
+	 * @param array $configs Full sync configuration for all sync modules.
+	 */
+	public function continue_enqueuing_with_lock( $configs = null ) {
 		if ( ! $configs ) {
 			$configs = $this->get_config();
 		}
 
-		if ( ! $enqueue_status ) {
-			$enqueue_status = $this->get_enqueue_status();
-		}
-
-		$modules           = Modules::get_modules();
-		$modules_processed = 0;
-		foreach ( $modules as $module ) {
-			$module_name = $module->name();
-
-			// Skip module if not configured for this sync or module is done.
-			if ( ! isset( $configs[ $module_name ] )
-				|| // No module config.
-					! $configs[ $module_name ]
-				|| // No enqueue status.
-					! $enqueue_status[ $module_name ]
-				|| // Finished enqueuing this module.
-					true === $enqueue_status[ $module_name ][2] ) {
-				$modules_processed ++;
-				continue;
-			}
-
-			list( $items_enqueued, $next_enqueue_state ) = $module->enqueue_full_sync_actions( $configs[ $module_name ], $remaining_items_to_enqueue, $enqueue_status[ $module_name ][2] );
-
-			$enqueue_status[ $module_name ][2] = $next_enqueue_state;
+		$this->remaining_items_to_enqueue = min( Settings::get_setting( 'max_enqueue_full_sync' ), $this->get_available_queue_slots() );
 
-			// If items were processed, subtract them from the limit.
-			if ( ! is_null( $items_enqueued ) && $items_enqueued > 0 ) {
-				$enqueue_status[ $module_name ][1] += $items_enqueued;
-				$remaining_items_to_enqueue        -= $items_enqueued;
-			}
-
-			if ( true === $next_enqueue_state ) {
-				$modules_processed ++;
+		foreach ( $this->get_remaining_modules_to_enqueue( $configs ) as $module ) {
+			if ( 0 >= $this->remaining_items_to_enqueue ) {
+				return;
```

Comment:
Should this be `break`? Otherwise we'll never enqueue items for the remaining modules?

Reply:
we don't want to continue enqueuing mor modules at this point, as we have reached the items to enqueue limit.